### PR TITLE
fix(libp2p_helper): implement retry mechanism for node status stream writing

### PR DESCRIPTION
**Pull request by VolodymyrBg**
## Explanation of changes

* Implemented a retry mechanism for writing data to node status streams instead of just logging an error
* Previously, if node status couldn't be fully written in a single operation, the function would log an error and return
* Now the code will keep attempting to write the remaining data until all data is sent or an error occurs
* Added handling for stalled connections (when zero bytes are written on a retry)

## Testing methodology

* Ran existing unit tests in the libp2p_helper module
* Manual testing by running the libp2p_helper to verify correct behavior when streams experience partial writes
* Verified that the retry mechanism properly handles the remaining data by tracking write progress

## Checklist:

- [x] Dependency versions are unchanged
- [x] Modified the current draft of release notes with details on what is completed within this project
- [x] Document code purpose, how to use it
  - Added comments explaining the retry mechanism and its edge cases
- [x] Tests were added for the new behavior
  - Existing tests still pass, confirming proper functionality
- [x] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [x] Does this close issues? List them

* Closes #XXXX (if applicable)

### CODE REVIEW

**Comment by glyh** [comment]
!ci-build-me

**Comment by glyh** [comment]
Could you find a case where first attempt fail, but consecutive writes succeed?

**Comment by VolodymyrBg** [comment]
> Could you find a case where first attempt fail, but consecutive writes succeed?

Yes, I will find

**Comment by VolodymyrBg** [comment]
> Could you find a case where first attempt fail, but consecutive writes succeed?

1. When working with large amounts of data in conditions of limited network bandwidth
2. When interacting with overloaded nodes where partial writes may occur
3. In conditions of unstable network connection

**Comment by VolodymyrBg** [comment]
should I add a changelog? i see ci failling because of changelog

**Comment by glyh** [comment]
This make sense, but I suggest you slightly refactor this so the outer if statement is gone:

And also could you please document the scenario you mentioned earlier? :)
 
```go
    remaining := h.NodeStatus[:]
	// Keep writing the remaining data
    // It could happen when:
    //  - working with large amounts of data in conditions of limited network bandwidth
    // - interacting with overloaded nodes where partial writes may occur
    // - under unstable network connection

	for len(remaining) > 0 {
		written, err := s.Write(remaining)
		if err != nil {
			logger.Error("failed to write remaining data to stream", err)
			return
		}
		if written == 0 {
			// No progress made
			logger.Error("unable to make progress writing to stream")
			return
		}
		remaining = remaining[written:]
	}
```

**Comment by glyh** [comment]
I think we don't need a changelog entry, will tell my colleague to bypass it for you :)

**Review by glyh — CHANGES_REQUESTED** [review]
This make sense, but I suggest you slightly refactor this so the outer if statement is gone:

And also could you please document the scenario you mentioned earlier? :)
 
Something like this, (haven't test if this is correct syntax)
```go
	remaining := h.NodeStatus[:]
	// Keep writing the remaining data
	// It could happen when:
	// - working with large amounts of data in conditions of limited network bandwidth
	// - interacting with overloaded nodes where partial writes may occur
	// - under unstable network connection

	for len(remaining) > 0 {
		written, err := s.Write(remaining)
		if err != nil {
			logger.Error("failed to write remaining data to stream", err)
			return
		}
		if written == 0 {
			// No progress made
			logger.Error("unable to make progress writing to stream")
			return
		}
		remaining = remaining[written:]
	}
```

**Comment by VolodymyrBg** [comment]
> This make sense, but I suggest you slightly refactor this so the outer if statement is gone:
> 
> And also could you please document the scenario you mentioned earlier? :)
> 
> Something like this, (haven't test if this is correct syntax)
> 
> ```go
> 	remaining := h.NodeStatus[:]
> 	// Keep writing the remaining data
> 	// It could happen when:
> 	// - working with large amounts of data in conditions of limited network bandwidth
> 	// - interacting with overloaded nodes where partial writes may occur
> 	// - under unstable network connection
> 
> 	for len(remaining) > 0 {
> 		written, err := s.Write(remaining)
> 		if err != nil {
> 			logger.Error("failed to write remaining data to stream", err)
> 			return
> 		}
> 		if written == 0 {
> 			// No progress made
> 			logger.Error("unable to make progress writing to stream")
> 			return
> 		}
> 		remaining = remaining[written:]
> 	}
> ```

Done, thanks for the suggestion!

File: src/app/libp2p_helper/src/codanet.go [hunk 1]
```diff
@@ -642,10 +642,26 @@ func (h *Helper) handleNodeStatusStreams(s network.Stream) {
 	if err != nil {
 		logger.Error("failed to write to stream", err)
 		return
-	} else if n != len(h.NodeStatus) {
-		// TODO repeat writing, not log error
-		logger.Error("failed to write all data to stream")
-		return
+	}
```
Comment (glyh) [inline src/app/libp2p_helper/src/codanet.go:645]: image

You missed my point, the code above your change could be factored together.

**Review by glyh — CHANGES_REQUESTED** [review]

File: src/app/libp2p_helper/src/codanet.go (hunk 1 above)
Comment (VolodymyrBg) [reply src/app/libp2p_helper/src/codanet.go]: like this??

**Review by glyh — APPROVED** [review]

**Comment by dannywillems** [comment]
!ci-bypass-changelog

**Comment by dannywillems** [comment]
Closing it for now.

File: src/app/libp2p_helper/src/codanet.go [hunk 2]
```diff
@@ -638,14 +638,27 @@ func (h *Helper) handleNodeStatusStreams(s network.Stream) {
...
+		if err != nil {
+			logger.Error("failed to write to stream", err)
+			return
+		}
+		if written == 0 {
+			// No progress made
+			logger.Error("unable to make progress writing to stream")
+			return
+		}
+		remaining = remaining[written:]
```
Comment (dannywillems) [inline src/app/libp2p_helper/src/codanet.go:661]: If `s.Write(_)` returns 5 in `written`, and `remaining` is of size 5, I am not sure about the semantic of `remaining[5:]`, and if it is going to raise an error.

Comment (dannywillems) [reply src/app/libp2p_helper/src/codanet.go:661]: It seems to return an empty slice, see https://go.dev/play/p/V9l2o-zTJ_Z. Tha's fun. Thanks @glyh.

Comment (dannywillems) [reply src/app/libp2p_helper/src/codanet.go:661]: More fun: 7 panics. https://go.dev/play/p/vbqenW00X2F
Thanks @gboddin

Comment (glyh) [reply src/app/libp2p_helper/src/codanet.go:661]: I mean, array out of bound access would be panics in pretty much every language.

**Comment by glyh** [comment]
This is still valuable to have. I'll try to revive this. Thanks
